File: workbook/ch07/sec7.6/patterns/strategy/sort/sort.c
#include <stdio.h>
#include <stdlib.h>

// Strategy Interface (Function Pointer Type)
typedef void (*SortFunction)(int*, size_t);

// Concrete Strategies
void bubble_sort(int* arr, size_t size) {
    for (size_t i = 0; i < size - 1; ++i) {
        for (size_t j = 0; j < size - i - 1; ++j) {
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

void quick_sort(int* arr, size_t size) {
    if (size < 2) return;
    int pivot = arr[size / 2];
    size_t i = 0, j = size - 1;

    while (i <= j) {
        while (arr[i] < pivot) i++;
        while (arr[j] > pivot) j--;
        if (i <= j) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            i++;
            j--;
        }
    }

    quick_sort(arr, j + 1);
    quick_sort(arr + i, size - i);
}

// Context
typedef struct {
    SortFunction sort_func;
} Sorter;

void sorter_init(Sorter* sorter, SortFunction func) {
    sorter->sort_func = func;
}

void sorter_perform_sort(Sorter* sorter, int* arr, size_t size) {
    sorter->sort_func(arr, size);
}

// Client Code
int main() {
    int data[] = {3, 1, 4, 1, 5, 9, 2, 6};
    size_t size = sizeof(data) / sizeof(data[0]);

    Sorter sorter;

    // Use Bubble Sort
    sorter_init(&sorter, bubble_sort);
    sorter_perform_sort(&sorter, data, size);
    printf("Bubble Sort: ");
    for (size_t i = 0; i < size; i++) printf("%d ", data[i]);
    printf("\n");

    // Reset data (since it was modified in-place)
    int reset_data[] = {3, 1, 4, 1, 5, 9, 2, 6};
    for (size_t i = 0; i < size; i++) data[i] = reset_data[i];

    // Switch to Quick Sort
    sorter_init(&sorter, quick_sort);
    sorter_perform_sort(&sorter, data, size);
    printf("Quick Sort:  ");
    for (size_t i = 0; i < size; i++) printf("%d ", data[i]);
    printf("\n");

    return 0;
}

File: workbook/ch07/sec7.6/patterns/strategy/sort/sort.py
from abc import ABC, abstractmethod

# Strategy Interface
class SortingStrategy(ABC):
    @abstractmethod
    def sort(self, data):
        pass

# Concrete Strategies
class BubbleSortStrategy(SortingStrategy):
    def sort(self, data):
        # Create a copy to avoid modifying the original
        sorted_data = list(data)
        n = len(sorted_data)
        for i in range(n):
            for j in range(n - i - 1):
                if sorted_data[j] > sorted_data[j + 1]:
                    sorted_data[j], sorted_data[j + 1] = sorted_data[j + 1], sorted_data[j]
        return sorted_data

class QuickSortStrategy(SortingStrategy):
    def sort(self, data):
        if len(data) <= 1:
            return list(data)
        pivot = data[len(data) // 2]
        left = [x for x in data if x < pivot]
        middle = [x for x in data if x == pivot]
        right = [x for x in data if x > pivot]
        return self.sort(left) + middle + self.sort(right)

# Context
class Sorter:
    def __init__(self, strategy):
        self.strategy = strategy

    def sort(self, data):
        return self.strategy.sort(data)

# Client Code
if __name__ == "__main__":
    data = [3, 1, 4, 1, 5, 9, 2, 6]

    # Use Bubble Sort
    sorter = Sorter(BubbleSortStrategy())
    sorted_data = sorter.sort(data)
    print("Bubble Sort:", sorted_data)

    # Switch to Quick Sort
    sorter.strategy = QuickSortStrategy()
    sorted_data = sorter.sort(data)
    print("Quick Sort:", sorted_data)